File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var countComponents = function(n, edges) {
    const adjList = new Array(n).fill().map(() => [])
    const visited = new Array(n).fill(false)

    for(const [start, end] of edges){
        adjList[start].push(end)
        adjList[end].push(start)
    }

    function helper(node){
        visited[node] = true
        for(const nei of adjList[node]){
            if(!visited[nei]){
                helper(nei)
            }
        }
    }

    let components = 0
    for(let i = 0; i < n; i++){
        if(!visited[i]){
            helper(i)
            components++
        }
    }
    return components
};

File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adjList = defaultdict(list)
        for a, b in edges:
            adjList[a].append(b)
            adjList[b].append(a)
        
        visited = set()
        def helper(node):
            q = deque([node])
            while q:
                n = q.popleft()
                visited.add(n)
                for neigh in adjList[n]:
                    if neigh not in visited:
                        q.append(neigh)
        
        count = 0
        for i in range(n):
            if i not in visited:
                helper(i)
                count += 1
        
        return count